source/extensions/router/cluster_specifiers/matcher/matcher_cluster_specifier.cc

```diff
@@ -0,0 +1,62 @@
+#include "source/extensions/router/cluster_specifiers/matcher/matcher_cluster_specifier.h"
+
+#include "envoy/extensions/router/cluster_specifiers/matcher/v3/matcher.pb.validate.h"
+
+#include "source/common/router/config_impl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Router {
+namespace Matcher {
+
+Envoy::Matcher::ActionFactoryCb ClusterActionFactory::createActionFactoryCb(
+    const Protobuf::Message& config, ClusterActionContext&,
+    ProtobufMessage::ValidationVisitor& validation_visitor) {
+  const auto& proto_config =
+      MessageUtil::downcastAndValidate<const ClusterActionProto&>(config, validation_visitor);
+  auto cluster = std::make_shared<std::string>(proto_config.cluster());
+
+  return [cluster]() { return std::make_unique<ClusterAction>(cluster); };
+}
+
+REGISTER_FACTORY(ClusterActionFactory, Envoy::Matcher::ActionFactory<ClusterActionContext>);
+
+class MatcherRouteEntry : public Envoy::Router::RouteEntryImplBase::DynamicRouteEntry {
+public:
+  MatcherRouteEntry(Envoy::Router::RouteEntryAndRouteConstSharedPtr parent,
+                    Envoy::Matcher::MatchTreeSharedPtr<Http::HttpMatchingData> match_tree)
+      : DynamicRouteEntry(std::move(parent), ""), match_tree_(std::move(match_tree)) {}
+
+  void refreshRouteCluster(const Http::RequestHeaderMap& headers,
+                           const StreamInfo::StreamInfo& stream_info) const {
+    Http::Matching::HttpMatchingDataImpl data(stream_info);
+    data.onRequestHeaders(headers);
+
+    Envoy::Matcher::MatchResult match_result =
+        Envoy::Matcher::evaluateMatch<Http::HttpMatchingData>(*match_tree_, data);
+
+    if (!match_result.isMatch()) {
+      return;
+    }
+
+    const Envoy::Matcher::ActionPtr result = match_result.action();
+    cluster_name_ = std::string(result->getTyped<ClusterAction>().cluster());
```

Comment:
Is it possible to avoid string copy?

```suggestion
    const auto& cluster_action = result->getTyped<ClusterAction>();
    cluster_name_ = cluster_action.cluster();
```

Reply:
It's possible but not be recommend for know. The dynamic route entry will always need to store cluster_name_; I will prefer to do a independent refactor to avoid inherit from DynamicRouteEntry and to avoid the unncessary string copy.